File: create_table/create-dict.py
"""
コミュニティのファイルから、それぞれのコミュニティごとの許可リストを作成するスクリプト
ここでは、自分のコミュニティ以外のノードからのアクセスをすべて許可すると仮定して行う

"""

import os

INPUT_DIR = "./../calc-modularity/new-new-community/"
OUTPUT_DIR = "./fb-new-table"
# INPUT_DIR = "./../../Louvain/community"
# OUTPUT_DIR = "./table"


# コミュニティファイルのリスト
# community_file_list = [
#     "ca-grqc-connected.cm",
#     "cmu.cm",
#     # "com-amazon-connected.cm",
#     # "email-enron-connected.cm",
#     "fb-caltech-connected.cm",
#     # "fb-pages-company.cm",
#     "karate-graph.cm",
#     "karate.cm",
#     "rt-retweet.cm",
#     "simple_graph.cm",
#     # "soc-slashdot.cm",
#     "tmp.cm",
# ]
community_file_list = [
    "2_communities.txt",  # 0
    "3_communities.txt",  # 0
    "4_communities.txt",  # 1
    "5_communities.txt",  # 2
    "6_communities.txt",  # 3
    "7_communities.txt",  # 4
    "8_communities.txt",  # 5
    "9_communities.txt",  # 6
    "10_communities.txt",  # 7
    "11_communities.txt",  # 8
    "12_communities.txt",  # 9
    "13_communities.txt",  # 10
    "14_communities.txt",  # 11
    "15_communities.txt",  # 12
    "16_communities.txt",  # 13
    "17_communities.txt",  # 14
    "18_communities.txt",  # 15
]


def select_graph():
    # ユーザーに選択肢を表示
    print("グラフファイルを選択してください:")
    for i, file_name in enumerate(community_file_list):
        print(f"{i}: {file_name}")

    # ユーザーからの入力を受け取る
    try:
        choice = int(input("番号を入力してください: "))
        if choice < 0 or choice >= len(community_file_list):
            raise ValueError("無効な番号が入力されました。")
        return community_file_list[choice]
    except ValueError as e:
        print(f"エラー: {e}")
        return None


def process_communities(graph_file_name):
    input_file_path = os.path.join(INPUT_DIR, graph_file_name)
    print(f"処理対象のファイル: {input_file_path}")

    # コミュニティごとにノードを格納するためのリストを初期化
    community_list = []

    # ファイルを読み込み、コミュニティごとにノードを分類
    with open(input_file_path, "r") as f:
        for line in f:
            node, community = map(int, line.split())
            # 必要ならば、リストを拡張して格納
            while len(community_list) <= community:
                community_list.append([])
            community_list[community].append(node)

    print("aaa")
    # 各コミュニティに対して処理を行う
    for community, key_nodes in enumerate(community_list):
        if not key_nodes:
            continue

        # 全ノードとそのコミュニティをリストに格納
        node_community_list = []
        with open(input_file_path, "r") as f:
            for line in f:
                node, community_num = map(int, line.split())
                node_community_list.append((node, community_num))

        # 結果を二重リスト形式で格納
        result_list = []
        for key_node in key_nodes:
            key_community = next(
                community for node, community in node_community_list if node == key_node
            )
            # 自分と同じコミュニティのノードと自身を除くノードをリストに格納
            value_nodes = [
                node
                for node, community_num in node_community_list
                if node != key_node and community_num != key_community
            ]
            result_list.append([key_node, value_nodes])

        # 指定の形式に変換する
        formatted_list = []
        for key_node, value_nodes in result_list:
            formatted_string = f"{key_node}: " + ", ".join(map(str, value_nodes))
            formatted_list.append(formatted_string)

        # 結果をファイルに書き出し
        graph_name = os.path.splitext(graph_file_name)[0]
        # output_dir = os.path.join("./new-community-table/", graph_name)
        output_dir = os.path.join(OUTPUT_DIR, graph_name)
        print(output_dir)
        os.makedirs(output_dir, exist_ok=True)  # 出力ディレクトリが存在しない場合は作成
        output_file_path = os.path.join(output_dir, f"community_{community}_result.txt")
        with open(output_file_path, "w") as out_file:
            for result in formatted_list:
                out_file.write(f"{result},\n")
        print(
            f"コミュニティ {community} の結果を {output_file_path} に書き出しました。"
        )


def main():
    # グラフを選択
    selected_graph = select_graph()

    for i in range(len(community_file_list)):
        process_communities(community_file_list[i])

    # if selected_graph:
    #     print(f"選択されたグラフ: {selected_graph}")
    #     process_communities(selected_graph)
    # else:
    #     print("グラフが選択されませんでした。")


if __name__ == "__main__":
    main()

File: algo/create_table/create-dict.cpp
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>
#include <sys/types.h>

#define MAX_COMMUNITIES 1000
#define MAX_NODES 10000

const char* input_dir = "./../../Louvain/community";
const char* output_dir = "./table";

const char* community_file_list[] = {
    "ca-grqc-connected.cm",
    "cmu.cm",
    "com-amazon-connected.cm",
    "fb-caltech-connected.cm",
    "karate-graph.cm",
    "karate.cm",
    "rt-retweet.cm",
    "simple_graph.cm",
    "tmp.cm",
};

int community_list[MAX_COMMUNITIES][MAX_NODES];
int community_size[MAX_COMMUNITIES];
int total_communities = 0;

void select_graph(char* selected_graph) {
    // グラフファイルを選択する
    int choice;
    printf("グラフファイルを選択してください:\n");
    for (int i = 0; i < sizeof(community_file_list) / sizeof(char*); i++) {
        printf("%d: %s\n", i, community_file_list[i]);
    }

    printf("番号を入力してください: ");
    scanf("%d", &choice);

    if (choice < 0 || choice >= sizeof(community_file_list) / sizeof(char*)) {
        printf("無効な番号が入力されました。\n");
        exit(1);
    }

    strcpy(selected_graph, community_file_list[choice]);
}

void process_communities(const char* graph_file_name) {
    char input_file_path[256];
    snprintf(input_file_path, sizeof(input_file_path), "%s/%s", input_dir, graph_file_name);

    FILE* fp = fopen(input_file_path, "r");
    if (fp == NULL) {
        perror("ファイルを開けませんでした");
        exit(1);
    }

    // コミュニティごとにノードを格納する
    int node, community;
    while (fscanf(fp, "%d %d", &node, &community) != EOF) {
        if (community >= MAX_COMMUNITIES) {
            printf("コミュニティ数が最大値を超えました。\n");
            exit(1);
        }
        community_list[community][community_size[community]++] = node;
    }

    fclose(fp);

    // 各コミュニティに対して処理を行う
    for (community = 0; community < MAX_COMMUNITIES; community++) {
        if (community_size[community] == 0 || community < 21) {
            continue;
        }

        // 出力ディレクトリの作成
        char output_dir_path[256];
        snprintf(output_dir_path, sizeof(output_dir_path), "%s/%s", output_dir, graph_file_name);
        mkdir(output_dir_path, 0755);

        // 結果をファイルに書き出し
        char output_file_path[256];
        snprintf(output_file_path, sizeof(output_file_path), "%s/community_%d_result.txt", output_dir_path, community);
        FILE* out_fp = fopen(output_file_path, "w");
        if (out_fp == NULL) {
            perror("出力ファイルを開けませんでした");
            exit(1);
        }

        // ノードリストの書き出し
        for (int i = 0; i < community_size[community]; i++) {
            int key_node = community_list[community][i];
            fprintf(out_fp, "%d: ", key_node);
            for (int c = 0; c < MAX_COMMUNITIES; c++) {
                if (c == community) {
                    continue;
                }
                for (int j = 0; j < community_size[c]; j++) {
                    fprintf(out_fp, "%d", community_list[c][j]);
                    if (j < community_size[c] - 1) {
                        fprintf(out_fp, ", ");
                    }
                }
            }
            fprintf(out_fp, "\n");
        }

        fclose(out_fp);
        printf("コミュニティ %d の結果を %s に書き出しました。\n", community, output_file_path);
    }
}

int main() {
    char selected_graph[256];
    select_graph(selected_graph);
    printf("選択されたグラフ: %s\n", selected_graph);
    process_communities(selected_graph);
    return 0;
}
